kyaml/setters2/settersutil/substitutioncreator.go

```diff
@@ -62,3 +73,114 @@ func (c SubstitutionCreator) Create(openAPIPath, resourcesPath string) error {
 		Outputs: []kio.Writer{inout},
 	}.Execute()
 }
+
+// CreateSettersForSubstitution creates the setters for all the references in the substitution
+// values if they don't already exist in openAPIPath file.
+func (c SubstitutionCreator) CreateSettersForSubstitution(openAPIPath string) error {
+	b, err := ioutil.ReadFile(openAPIPath)
+	if err != nil {
+		return err
+	}
+
+	// parse the yaml file
+	y, err := yaml.Parse(string(b))
+	if err != nil {
+		return err
+	}
+
+	m, err := c.GetValuesForMarkers()
+	if err != nil {
+		return err
+	}
+
+	// for each ref in values, check if the setter already exists, if not create them
+	for _, value := range c.Values {
+		obj, err := y.Pipe(yaml.Lookup(
+			// get the setter key from ref. Ex: from #/definitions/io.k8s.cli.setters.image_setter
+			// extract io.k8s.cli.setters.image_setter
+			"openAPI", "definitions", strings.TrimPrefix(value.Ref, "#/definitions/")))
+
+		if err != nil {
+			return err
+		}
+
+		if obj == nil {
+			sd := setters2.SetterDefinition{
+				// get the setter name from ref. Ex: from #/definitions/io.k8s.cli.setters.image_setter
+				// extract image_setter
+				Name:        strings.Split(value.Ref, ".")[4],
+				Value:       m[value.Marker],
+				Description: "Created for substitution " + c.Name,
+			}
+			sd.AddToFile(openAPIPath)
+		}
+	}
+	return nil
+}
+
+// GetValuesForMarkers parses the pattern and field value to derive values for the
+// markers in the pattern string. Returns error if the marker values can't be derived
+func (c SubstitutionCreator) GetValuesForMarkers() (map[string]string, error) {
+	m := make(map[string]string)
+	inds, err := c.GetStartIndices()
+	if err != nil {
+		return nil, err
+	}
+	s:=c.FieldValue
+	p:=c.Pattern
+	i:=0
+	j:=0
+	// iterate s, p with indices i, j respectively and when j hits the index of a marker,
+	// freeze j and iterate i till we find the character just after marker and capture the string
```

Comment:
don't we need to match the substring after the marker, not just the character?  what if the character after the marker is part of the value -- e.g. pattern `/projects/PROJECT_SETTER/zones/ZONE_SETTER` with value `/projects/project1//zones/zone1` -- this would stop the capture when it its the first `/`, but should stop when it hits the second

Reply:
Done